### ISSUE TITLE
Automatic Export not working

### ISSUE THREAD

**Issue by owenhdiba** [retorquere/zotero-better-bibtex#1958]
Trying to get automatic exports working for bibtex files. When I export a collection and tick the "keep updated" box occasionally an error message will pop up informing me that it failed to start background exports. Sometimes no message comes up. In both cases, zotero fails to update the file when i make changes to the collection. If i navigate to the "automatic export" bar of the BBT settings and manually request that the file is updated it does successfully update the file. Running Mac OS 11.6 and have ztoero fully updated. Have tried reinstalling both zotero and BBT to no avail.

**Comment by retorquere** [comment]
There isn't anything in `HG3DU239-euc` except messages about your storage quota having run out. Did you experience the problem again? What scenario did you go through that gave you `HG3DU239-euc`?

**Comment by vin047** [comment]
I'm having the same issue as @owenhdiba. After some testing, i've discovered that it occurs only after the Zotero window is closed; if you don't close the Zotero window, auto export works fine. It stays broken even if you reopen the window later – you have to restart Zotero to get auto export working again. Also the issue only occurs after the Zotero window has been closed for some time (unsure how long exactly but at least 30 mins, possibly sooner). And it occurs regardless of whether debug logs are enabled or not.

I tried generating the Better BibTeX debug report but got this error message:
`[Exception... "Component not initialized"  nsresult: "0xc1f30001 (NS_ERROR_NOT_INITIALIZED)"  location: "JS frame :: chrome://zotero-better-bibtex/content/better-bibtex.js :: node_modules/jszip/dist/jszip.min.js/</<[54]</e.exports< :: line 72751"  data: no] (MVZVF4PR-euc, references: false)`

Not sure if it's because I use Firefox, not Chrome...

So I tried following your advice here:
> You can start Zotero using
> 
> ```
> /Applications/Zotero.app/Contents/MacOS/zotero -ZoteroDebugText > ~/1958.log &
> ```
> 
> and it will log everything it does in `~/1958.log`; you don't specifically need to keep Zotero running, if you just start Zotero this way and you experience the problem, you can upload `1958.log` (or whatever you choose for the log file name).

I ran two tests in debug mode:
1. saving a webpage sometime after starting Zotero, without closing the Zotero window.
2. saving a webpage sometime after starting Zotero, with the window having been closed during this time.

Each time, I started zotero with the command you specified, then after waiting period, navigated to the site in Firefox and saved it with the extension. (After each test, I quit Zotero, launched it normally, removed the saved site, then restarted with command you specified, to try and use the same site for both tests and run from similar starting state). Auto export worked in test 1 but not in test 2. Attached the debug log for both tests below. 

Used `https://www.nickseitz.com/writing/take-less-stupid-notes` as the test site.

[test1-site-saved-at-9.15pm.log](https://github.com/retorquere/zotero-better-bibtex/files/7523875/test1-site-saved-at-9.15pm.log)
[test2-site-saved-at-11.06pm.log](https://github.com/retorquere/zotero-better-bibtex/files/7523876/test2-site-saved-at-11.06pm.log)

**Comment by retorquere** [comment]
Thanks -- this means it likely has to do with the fact that MacOS programs keep running even when all of their windows have closed. This has caused problems before, maybe these logs can give me a hint on what to do.

This behavior of MacOS is such a pain. I'm on MacOS myself, in the balance I find it the most useful OS for me, but this is one of the parts I really dislike.

**Comment by champagnealexandre** [comment]
Having the same problem — I see the `awaiting-user-feedback` was added back to this issue, do you need any log/anything else @retorquere ?

**Comment by retorquere** [comment]
No, I just forgot to remove the tag that was automatically added. I'm working on fixing this, but it's slow going, I need to redo my bundling stack for it.

**Comment by champagnealexandre** [comment]
okay no worries, bbt is still functional and the workaround is to quit/reopen zotero so it’s not that critical

thanks!

**Comment by owenhdiba** [comment]
> There isn't anything in `HG3DU239-euc` except messages about your storage quota having run out. Did you experience the problem again? What scenario did you go through that gave you `HG3DU239-euc`?

Hi sorry for the slow reply. The scenario was: I added something to a collection with auto export enabled, noticed that it had failed and then generated the support log. I'm not sure why nothing showed up.

If it's helpful, i've noticed that if I right click on the collection with auto-export and select "automatic export" in the context menu it comes up with a little empty panel which seems odd. If I click it the bibtex file still doesn't update.

If i navigate to the zotero preferences and then auto-update from there it does update the file. Occasionally when i do this it does actually pop with an error message in the zotero window saying something like "background export failed to start after 5 attempts"

**Comment by retorquere** [comment]
Can each of you edit your first message here to add the operating system in use?

**Comment by tim-hilde** [comment]
Maybe to add to this, I have the same issues and I am using macOS 12.1. Hope this helps fixing the issue

**Comment by egrefen** [comment]
I have this issue on both OS X 10.15.7 (19H1606) on my home iMac, and OS X 12.1 (21C52) on my work mac. Both off fresh installs of Zotero (checkted that it was the latest version), with Better Bibtex up to date. The library is pulled via sync with my zotero account. Automatic export next works, although forcing it does (but throws up the following error box):
![Screenshot 2022-02-04 at 20 44 44](https://user-images.githubusercontent.com/1145770/152601285-4a2d228f-9bb0-43fa-8a54-734f7f0bf729.png)

**Comment by retorquere** [comment]
@egrefen you are experiencing a different error.

**Comment by egrefen** [comment]
> @egrefen you are experiencing a different error.

Sorry (although this is what I get when I try to force export after noting that automatic export doesn't work, which it doesn't on either mac). Is there another issue I should be following, or should I raise a new one, @retorquere?

**Comment by retorquere** [comment]
Yes please

**Comment by bfbraum** [comment]
I'm also experiencing a failure to export issue... sort of. I've been using automatic export for around a year now, on group libraries, so that my research team can add references to a group library, trigger Zotero on my desktop (late 2013 Mac Pro, currently running OS 12.4, which I leave running for this reason), and have an updated .bib file in our shared Overleaf directory, all usually in under 10 seconds. Slick system when it works, but I've just realized that it's not. I'm on Zotero 6.0.9, and I deleted and reinstalled BetterBibTeX, to no effect.

The weird thing, which I just noticed and confirmed as I was writing this, is that Zotero _does_ export the group library... eventually. Twice, now, I've updated the library and found that the .bib file is updated 8 minutes later. The first time, it was set to automatic export on change; the second time, when idle. Both updated 8 minutes after I made the change. So this might be a separate issue, or it might be that other people who have this problem aren't waiting 8 minutes. 

I generated a support log (for one of the attempts when I didn't wait 8 minutes): QXPJYSVP-euc.

**Comment by grahamannett** [comment]
I've been having this issue lately. was working and then stopped at some point and not entirely sure why. restarting zotero with better bibtex logging (from help menu) enabled and seems like it's fixed now. no errors listed in the debug log so not entirely sure how it could have fixed it but who knows.  

very odd but might help someone else to just enable logging and restart.

**Comment by laras126** [comment]
Thanks @grahamannett - my exported library was not updating when I added new citations, and enabling the debugging log from the help menu and it has started working now.

**Comment by choc96208** [comment]
The thing that worked for me was to reset `autoExportDelay` to the default of 5 (see Preferences > Advanced > Config Editor). Previously it was 567 and the export wasn't working. I found this solution at [Automatic export of Better BibTeX file is scheduled, but not executed](https://pullanswer.com/questions/automatic-export-of-better-bibtex-file-is-scheduled-but-not-executed).

**Comment by retorquere** [comment]
Export was working in your case. It was just delayed very long - by request. I can't help it if people set the delay to a long, non-standard duration.

**Comment by choc96208** [comment]
Yeah it probably was a separate issue. This was just one of search results I found. Thought if someone was having the same issue as me, it might be helpful. But I think I just spammed with an unrelated thing. Sorry about that.
